pkg/volume/flexvolume/driver-call.go

```diff
@@ -0,0 +1,230 @@
+	}
+}
+
+func (dc *DriverCall) Append(arg string) {
+	dc.args = append(dc.args, arg)
+}
+
+func (dc *DriverCall) AppendSpec(spec *volume.Spec, host volume.VolumeHost, extraOptions map[string]string) error {
+	optionsForDriver, err := NewOptionsForDriver(spec, host, extraOptions)
+	if err != nil {
+		return err
+	}
+
+	jsonBytes, err := json.Marshal(optionsForDriver)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal spec, error: %s", err.Error())
+	}
+
+	dc.Append(string(jsonBytes))
+	return nil
+}
+
+func (dc *DriverCall) Run() (*DriverStatus, error) {
+	if dc.plugin.isUnsupported(dc.Command) {
+		return nil, errors.New(StatusNotSupported)
+	}
+	execPath := dc.plugin.getExecutable()
+
+	cmd := dc.plugin.runner.Command(execPath, dc.args...)
+
+	timeout := false
+	if dc.Timeout > 0 {
+		timer := time.AfterFunc(dc.Timeout, func() {
+			timeout = true
+			cmd.Stop()
+		})
+		defer timer.Stop()
+	}
+
+	output, execErr := cmd.CombinedOutput()
+	if execErr != nil {
+		if timeout {
+			return nil, TimeoutError
+		}
+		glog.Errorf("FlexVolume: driver call failed: executable: %s, args: %s, error: %s, output: %s", execPath, dc.args, execErr.Error(), output)
+		_, err := handleCmdResponse(dc.Command, output)
+		if err == nil {
+			glog.Errorf("FlexVolume: driver bug: %s: exec error (%s) but no error in response.", execPath, execErr)
+			return nil, execErr
+		}
+		if isCmdNotSupportedErr(err) {
+			dc.plugin.unsupported(dc.Command)
+		}
+		return nil, err
+	}
+
+	status, err := handleCmdResponse(dc.Command, output)
+	if err != nil {
+		if isCmdNotSupportedErr(err) {
+			dc.plugin.unsupported(dc.Command)
+		}
+		return nil, err
+	}
+
+	return status, nil
+}
+
+// OptionsForDriver represents the spec given to the driver.
+type OptionsForDriver map[string]string
+
+func NewOptionsForDriver(spec *volume.Spec, host volume.VolumeHost, extraOptions map[string]string) (OptionsForDriver, error) {
+	volSource, readOnly := getVolumeSource(spec)
+	options := map[string]string{}
+
+	options[optionFSType] = volSource.FSType
+
+	if readOnly {
+		options[optionReadWrite] = "ro"
+	} else {
+		options[optionReadWrite] = "rw"
+	}
+
+	if volSource.SecretRef != nil {
+		// Extract secret and pass it as options.
+		secrets, err := getSecrets(spec, host)
+		if err != nil {
+			return nil, err
+		}
+		for name, secret := range secrets {
+			options[optionKeySecret+"/"+name] = secret
+		}
+	}
+
+	for key, value := range extraOptions {
+		options[key] = value
+	}
+
+	for key, value := range volSource.Options {
+		options[key] = value
+	}
+
+	return OptionsForDriver(options), nil
+}
+
+// DriverStatus represents the return value of the driver callout.
+type DriverStatus struct {
+	// Status of the callout. One of "Success", "Failure" or "Not supported".
+	Status string
+	// Message is the reason for failure.
+	Message string
+	// Device assigned by the driver.
+	Device string `json:"device"`
+	// The path asked by the plugin
+	Path string `json:"path"`
```

Comment:
We need to be very careful about what values are returned by what methods. Ideally we should have separate structs for different calls.

I think there are at least three different pieces of non-error information returned by the various calls:
```GO
// DevicePath is the path the device is attached.
// This field is returned by Attach calls and WaitForAttach calls.
// ie: /dev/sdx
DevicePath string `json:"device"`

// GlobalMountPath is the path where the device is (or should be) globally mounted.
// This field is returned by GetDeviceMountPath calls.
// ie: /var/lib/kubelet/plugins/kubernetes.io/flexvolume/myDriver/mounts/panda-disk/
GlobalMountPath string `json:"path"`

// UniqueVolumeName is the name/ID to uniquely identifying the actual backing
// device, directory, path, etc. referenced by the specified volume spec.
// For Attachable volumes, this value must be able to be passed back to volume Detach
// methods to identify the device to act on.
// This field is returned by GetDeviceMountPath calls.
// ie: /special/mount/path
UniqueVolumeName string `json:"path"`
```

`GlobalMountPath` actually shouldn't even be up to the driver. It should be auto-generated by flexvolume using the driver name and uniqueVolumeName.